src/Security/PasswordExpirationMiddleware.php

```diff
@@ -0,0 +1,265 @@
+        '*/*',
+        'text/*',
+        'text/html',
+        'application/xhtml+xml',
+        'text/xml',
+        'application/xml'
+    ];
+
+    public function process(HTTPRequest $request, callable $delegate)
+    {
+        try {
+            if ($response = $this->checkForExpiredPassword($request)) {
+                return $response;
+            }
+        } catch (DatabaseException $e) {
+            // Database isn't ready, carry on.
+        }
+
+        return $delegate($request);
+    }
+
+    /**
+     * Check if the just authenticated member has the password expired.
+     * Returns a response if the current request should not be
+     * processed as usual.
+     *
+     * @param HTTPRequest $request
+     *
+     * @return HTTPResponse|null
+     */
+    protected function checkForExpiredPassword(HTTPRequest $request): ?HTTPResponse
+    {
+        $session = $request->getSession();
+
+        if ($session && $session->get(static::SESSION_KEY_ALLOW_CURRENT_REQUEST)) {
+            // allow current request and skip the expiration check, but for only the current
+            // request, so we're deleting the flag from the session so it's not affecting other
+            // requests.
+            // This flag would usually be set from within $handler->authenticateRequest()
+            $session->clear(static::SESSION_KEY_ALLOW_CURRENT_REQUEST);
+
+            return null;
+        }
+
+        $user = Security::getCurrentUser();
+        if ($user && $user->isPasswordExpired()) {
+            if ($response = $this->handleExpiredPassword($request)) {
+                return $response;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Check if we have a redirect to a password change form registered
+     * and redirect there if possible.
+     * Otherwise, deauthenticate the user by resetting it for this request,
+     * since we should treat ones with expired passwords as unauthorised.
+     *
+     * @param HTTPRequest $request
+     *
+     * @return HTTPResponse|null
+     */
+    protected function handleExpiredPassword(HTTPRequest $request): ?HTTPResponse
+    {
+        $session = $request->getSession();
+
+        $sessionRedirectUrl = $session->get(static::SESSION_KEY_REDIRECT);
+        $defaultRedirectUrl = static::config()->get('default_redirect');
+
+        if ($sessionRedirectUrl || $defaultRedirectUrl) {
+            $redirectUrl = $this->absoluteUrl($sessionRedirectUrl ?? $defaultRedirectUrl);
+        } else {
+            $redirectUrl = null;
+        }
+
+        if (!$session || !$redirectUrl) {
+            Security::setCurrentUser(null);
+            return null;
+        }
+
+        $whitelist = [];
+
+        $currentUrl = $this->absoluteUrl($request->getURL(true));
+
+        $allowedUrls = array_merge([
+            $redirectUrl
+        ], array_map(
+            function ($url) {
+                return $this->absoluteUrl($url);
+            },
+            $whitelist
```

Comment:
array_map is being called on an empty array here isn't it?

Reply:
yes, a leftover after some deleted functionality. Will fix it :+1: